### DISCUSSION TITLE
Option to disable hashing for SSG

### DISCUSSION THREAD

**Discussion by jekuer** [Baroshem/nuxt-security#215] [Q&A]
Hi!

Is there a way to disable the hashing of JavaScript?
I am using some CSP and Nuxt SSG.
It would work fine, since I added "unsafe-inline". However, "unsafe-inline" is not respected, because this module adds some hashes to the CSP. I do not need the hashed values with unsafe-inline. Can I turn this off somehow?

Thanks in advance!
Jens

**Comment by Baroshem** [comment]
Hey Jens,

I am afraid that I do not understand fully what you would like to achieve. Could you please create some reproduction so that I could test it out?

I would like to help but at this point I do not understand fully what should be done :)

**Reply by jekuer** [reply]
Hi Jakub,

Thanks for you offer!

I am struggeling a little bit with setting up a minimal demo version.
But you can have a look at https://add-to-calendar-button.com, which is build from https://github.com/add2cal/add-to-calendar-button/tree/main/demo.
Nuxt config can be found at https://github.com/add2cal/add-to-calendar-button/blob/main/demo/nuxt.config.ts.

When inspecting the page, you should see the following code from the plugin:

```
<meta http-equiv="Content-Security-Policy" content="base-uri 'self'; font-src 'self' data:; form-action 'self'; img-src 'self' https://add-to-calendar-button.com data:; object-src 'none'; script-src-attr 'self'; style-src 'self' 'unsafe-inline' https://add-to-calendar-button.com; upgrade-insecure-requests ; script-src 'self' 'unsafe-inline' https://*.add-to-calendar-button.com 'sha256-UMS0FTau8dt6HxrSHR62fubYrv4p92+gJBT6tA8/2a4=' 'sha256-wGX5Vp+k5wB1wakIv0aUlYfRP9EPV8TA3/q0UPpW/js='">
```

I am wondering how I can prevent those hashes ("sha256-*") from being generated?

It would be also fine for me to disabled the CSP feature, as I set it on server side anyway. But if I do not specify it at all, the plugin generates some defaults, which is even worse.

**Reply by Baroshem** [reply]
You could try to disable nonce like following:

```ts
  security: {
    nonce: false,
  }
```

Or other options here https://nuxt-security.vercel.app/security/headers#nonce-support

**Reply by Baroshem** [reply]
More context https://github.com/Baroshem/nuxt-security/issues/178#issuecomment-1639859066

**Reply by jekuer** [reply]
This has nothing to do with the nonce feature.
It is basically in 02-cspSsg.ts line 47:
`html.head.push(generateCspMetaTag(contentSecurityPolicies, scriptHashes))`

The scriptHashes get generated and added no matter the other settings.
There should be some option to disable scriptHashes.
OR (at least) an option to disable CSP. As mentioned at many other places (also the linked discussion), for SSG sites, CSP is not that helpful. It should be possible to disable it somehow. Either via an explicit option (like with nonce) - or implicitly, if, for example there is an empty object set. The explicit options should be easier and cleaner.
Like adding a line to `isContentSecurityPolicyEnabled` saying something like:
```
if (ModuleOptions.csp === false) {
  return false;
}
```

**I could open a PR, if this is something you want to add?**

**Reply by Baroshem** [reply]
Ok, I think I get it now. 

Please correct me if I am wrong here but I think that what you are trying to achieve here is to be able to disable CSP generation on SSG while still having it on SSR (or at least have an option to disable script hashes). Is that correct?

Because you can right now disable a header completely if you like but this will also stop generating it on the SSR.

**Reply by jekuer** [reply]
Yes, exactly 😃 .
This would be the ideal setup.
However, disabling CSP completely would be also fine for now - how can I achieve this? If I simply do not declare it, it falls back to the defaults.

**Reply by Baroshem** [reply]
I just inspected the the code of your deployed application and the repository and there must be a bug somewhere. For some reason, in your app, the csp is being added to the head / meta even though it should only be applied on SSG pages.

I have tested it on a few application examples and when I was generating a SSG app with yarn generate and then yarn preview, the app was a SSG and it had the CSP in the head/meta while for the ones that I build with yarn build for SSR, the CSP was configured as a response header only.

**Reply by Baroshem** [reply]
This is the generated head/meta with your nuxt config in my app:

```html
<head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="modulepreload" as="script" crossorigin="" href="/_nuxt/entry.e144b3a9.js">
<link rel="modulepreload" as="script" crossorigin="" href="/_nuxt/index.495b5b16.js">
<link rel="prefetch" as="script" crossorigin="" href="/_nuxt/error-404.c3e43f45.js">
<link rel="prefetch" as="script" crossorigin="" href="/_nuxt/vue.f36acd1f.fea4355f.js">
<link rel="prefetch" as="script" crossorigin="" href="/_nuxt/_plugin-vue_export-helper.c27b6911.js">
<link rel="prefetch" as="script" crossorigin="" href="/_nuxt/error-500.a532116d.js">
<script type="module" src="/_nuxt/entry.e144b3a9.js" crossorigin=""></script></head>
```

**Reply by jekuer** [reply]
My app is an SSG app, deployed as static site.

**Reply by Baroshem** [reply]
![image](https://github.com/Baroshem/nuxt-security/assets/37120330/6203f4eb-68ef-4141-b934-abdf19bdead4)

```ts
ssr: true,
```

**Reply by jekuer** [reply]
Yes, as per Nuxt documentation (https://nuxt.com/docs/getting-started/deployment), this is required for SSG.
Mind that I am refering to static-site-generation (SSG), not server-side-rendering (SSR)!

But, i have the same behavior at another app, where I set `ssr: false` 😉

**Reply by Baroshem** [reply]
I am trying to figure out what is causing this.

It seems that your app is some sort of a mix between SSG and SSR (because meta should be in SSG, and headers should be in SSR when in your case you get both for some reason).

You can always disable the header like this:

```ts
security: {
   headers: {
    contentSecurityPolicy: false
   }
}
```

But this may also disable generating this header on the server.

**Reply by jekuer** [reply]
Ah, I understand your confusion!
The headers you see in the response are coming from the server and are generated independently of this module!

**Reply by Baroshem** [reply]
This explains a lot. I thought that this was caused by the module.

Then if you dont want CSP for SSG (and you have it in one application) you can use the rule I showed above. It will not register the nitro plugin and the meta wont be added to your application.

**Reply by Baroshem** [reply]
However, for disabling the hashes, I think we would need to add another configuration option like:

```ts
security: {
   ...
   ssg: {
      hashScripts: false
   }
} 
```

If you would like to contribute with such idea, please do so for 1.0.0-rc.1 branch.

**Reply by jekuer** [reply]
Awesome!
I missed the option to disable a single header. Working now.
Sorry for the confusion. Really appreciate your active support!!

Regarding the hashes: Any reason why you prefer it nested below "ssg"? Would make this rather more global.

**Reply by Baroshem** [reply]
Because it is only used in SSG. For SSR apps, user will use the response headers. And also to keep it organized in the module configuration as other options usually refer to headers or middleware

**Reply by Baroshem** [reply]
Added this functionality here https://github.com/Baroshem/nuxt-security/pull/212/commits/644deee0a1687dfeec531ac960e4a85e47d6f5aa :)
